diagnostics/diagnostics_manager.rs

```diff
@@ -0,0 +1,155 @@
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::{collections::HashSet, future::Future, hash::Hash, sync::Arc};
+
+use error::TypeDBError;
+use resource::constants::{database::INTERNAL_DATABASE_PREFIX, diagnostics::REPORTING_URI};
+use tonic::Status;
+use tonic_types::StatusExt;
+
+use crate::{
+    metrics::{ActionKind, DatabaseMetrics, LoadKind},
+    monitoring_server::MonitoringServer,
+    reporter::Reporter,
+    Diagnostics,
+};
+
+macro_rules! diagnostics_method {
+    ($(
+        pub fn $fn_name:ident(&self, $($arg_name:ident: $arg_type:ty),* $(,)?) ;
+    )*) => {
+        $(
+            pub fn $fn_name(&self, $($arg_name: $arg_type),*) {
+                self.diagnostics.$fn_name($($arg_name),*)
+            }
+        )*
+    };
+}
+
+#[derive(Debug)]
+pub struct DiagnosticsManager {
+    diagnostics: Arc<Diagnostics>,
+    reporter: Option<Reporter>,
+    monitoring_server: Option<MonitoringServer>,
+}
+
+impl DiagnosticsManager {
+    pub fn new(
+        diagnostics: Diagnostics,
+        monitoring_port: u16,
+        is_monitoring_enabled: bool,
+        is_development_mode: bool,
+    ) -> Self {
+        let deployment_id = diagnostics.server_properties.deployment_id().clone().to_owned();
+        let data_directory = diagnostics.server_metrics.data_directory().clone();
+        let is_reporting_enabled = diagnostics.server_properties.is_reporting_enabled();
+        let diagnostics = Arc::new(diagnostics);
+
+        let reporter = if is_development_mode {
+            None
+        } else {
+            Some(Reporter::new(deployment_id, diagnostics.clone(), REPORTING_URI, data_directory, is_reporting_enabled))
+        };
+
+        let monitoring_server = if is_monitoring_enabled {
+            Some(MonitoringServer::new(diagnostics.clone(), monitoring_port))
+        } else {
+            None
+        };
+
+        Self { diagnostics, reporter, monitoring_server }
+    }
+
+    diagnostics_method! {
+        pub fn submit_database_metrics(&self, database_metrics: HashSet<DatabaseMetrics>);
+        pub fn submit_error(&self, database_name: Option<impl AsRef<str> + Hash>, error_code: String);
+        pub fn submit_action_success(&self, database_name: Option<impl AsRef<str> + Hash>, action_kind: ActionKind);
+        pub fn submit_action_fail(&self, database_name: Option<impl AsRef<str> + Hash>, action_kind: ActionKind);
+        pub fn increment_load_count(&self, database_name: &str, connection_: LoadKind);
+        pub fn decrement_load_count(&self, database_name: &str, connection_: LoadKind);
+    }
+
+    pub fn may_start_reporting(&self) {
+        if let Some(reporter) = &self.reporter {
+            reporter.may_start();
+        }
+    }
+
+    pub async fn may_start_monitoring(&self) {
+        if let Some(server) = &self.monitoring_server {
+            server.start_serving().await;
+        }
+    }
+}
+
+pub fn run_with_diagnostics<F, T>(
```

Comment:
These wrappers can be simply used when we perform a user action:
`run_with_diagnostics(...... { transaction.commit() })`
This way, we don't need to worry about counting actions, their results, and error codes. Everything is accounted.